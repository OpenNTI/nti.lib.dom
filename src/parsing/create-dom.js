import { applyStyles } from '../style/apply-styles';

const isConfig = RegExp.prototype.test.bind(
	/^(?:xmlns|tag|children|cn|html|style)$/i
);

/**
 * Creates new DOM element(s)
 *
 * @param {Object/String} o DOM spec
 * @param {HTMLElement} [parentNode] Optional DOM node to append the generated elements.
 * @param {string} xmlns
 * @returns {HTMLElement} The new node
 */
export function createDOM(o, parentNode, xmlns = o?.xmlns) {
	let el;
	let doc = parentNode ? parentNode.ownerDocument : document;

	if (!o) {
		return o;
	}

	if (Array.isArray(o)) {
		// Allow Arrays of siblings to be inserted
		el = doc.createDocumentFragment(); // in one shot using a DocumentFragment

		for (let i of o) {
			createDOM(i, el, xmlns);
		}
	} else if (typeof o === 'string') {
		el = doc.createTextNode(o);
	} else {
		if (xmlns) {
			el = doc.createElementNS(xmlns, o.tag);
		} else {
			el = doc.createElement(o.tag || 'div');
		}

		for (let attr in o) {
			if (!isConfig(attr)) {
				el.setAttributeNS(
					null,
					attr === 'cls' ? 'class' : attr,
					o[attr]
				);
			}
		}

		applyStyles(el, o.style);

		let cn = o.children || o.cn;
		if (cn) {
			createDOM(cn, el, xmlns);
		} else if (o.html) {
			el.innerHTML = o.html;
		}
	}

	if (parentNode) {
		parentNode.appendChild(el);
	}

	return el;
}
