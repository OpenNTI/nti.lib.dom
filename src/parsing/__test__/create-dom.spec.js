/* eslint-env jest */
import { createDOM } from '../create-dom';

test ('createDOM', () => {

	expect(() => createDOM()).not.toThrow();

	expect(createDOM({})).toMatchSnapshot();

	expect(createDOM({
		tag: 'span',
		cls: 'foobar',
		for: 'lala',
		cn: [
			{class: 'baz'},
			{cls: 'boo'},
			{html: 'hi'},
			'sup'
		]
	})).toMatchSnapshot();
});
