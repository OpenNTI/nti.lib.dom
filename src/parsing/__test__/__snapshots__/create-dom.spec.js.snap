// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createDOM 1`] = `<div />`;

exports[`createDOM 2`] = `
<span
  class="foobar"
  for="lala"
>
  <div
    class="baz"
  />
  <div
    class="boo"
  />
  <div>
    hi
  </div>
  sup
</span>
`;
